File: courses/codes/Bowling.java
import java.util.Scanner;
import java.util.Arrays;

public class Bowling {
    static int n, k, w;
    static int[][] dp = new int[501][30000];
    static int[] blocks = new int[30000];

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        int t = sc.nextInt();

        while (t-- > 0) {
            n = sc.nextInt();
            k = sc.nextInt();
            w = sc.nextInt();

            int[] pins = new int[n];
            int total = 0;

            for (int i = 0; i < n; i++) {
                pins[i] = sc.nextInt();
                total += pins[i];
            }

            // Precompute blocks (windows of size w)
            blocks[0] = 0;
            for (int i = 0; i < w; i++) {
                blocks[0] += pins[i];
            }

            for (int i = 1; i < n; i++) {
                blocks[i] = blocks[i - 1] - pins[i - 1];
                if (i + w - 1 < n) {
                    blocks[i] += pins[i + w - 1];
                }
            }

            // Reset dp
            for (int[] row : dp) {
                Arrays.fill(row, 0);
            }

            for (int i = 1; i <= k; i++) {
                for (int j = n - 1; j >= 0; j--) {
                    if (j >= n - w) {
                        dp[i][j] = blocks[j];
                    } else {
                        dp[i][j] = Math.max(dp[i - 1][j + w] + blocks[j], dp[i][j + 1]);
                    }
                }
            }

            System.out.println(dp[k][0]);
        }

        sc.close();
    }
}

File: courses/codes/cclv1-sr-class/bowling.py
MAX_K = 501
MAX_N = 30000

dp = [[0] * MAX_N for _ in range(MAX_K)]
blocks = [0] * MAX_N

def main():
    t = int(input())

    for _ in range(t):
        n, k, w = map(int, input().split())
        pins = [int(input()) for _ in range(n)]

        # Precompute blocks (window sums of size w)
        blocks[0] = sum(pins[:w])
        for i in range(1, n):
            blocks[i] = blocks[i - 1] - pins[i - 1]
            if i + w - 1 < n:
                blocks[i] += pins[i + w - 1]
            else:
                blocks[i] = 0  # Invalidate partial windows past the end

        # Reset dp for this test case
        for i in range(k + 1):
            for j in range(n):
                dp[i][j] = 0

        # Fill DP table
        for i in range(1, k + 1):
            for j in range(n - 1, -1, -1):
                if j >= n - w:
                    dp[i][j] = blocks[j]
                else:
                    dp[i][j] = max(dp[i - 1][j + w] + blocks[j], dp[i][j + 1])

        print(dp[k][0])

if __name__ == "__main__":
    main()